payments_redsys/test_redsys.py

```diff
@@ -0,0 +1,208 @@
+            self.redsys.shared_secret,
+        )
+
+        request = self.factory.get(
+            "/",
+            data={
+                "Ds_SignatureVersion": "HMAC_SHA256_V1",
+                "Ds_MerchantParameters": merchant_params,
+                "Ds_Signature": signature,
+            },
+        )
+
+        result = self.redsys.process_data(self.payment, request)
+
+        assert result.status_code == 302
+        assert result.url == f"http://localhost:8000/{self.payment.pk}/success"
+        self.payment.refresh_from_db()
+        assert self.payment.status == "confirmed"
+
+    @patch("payments_redsys.compare_signatures", Mock(return_value=True))
+    def test_process_data_rejected(self):
+        redsys_response = redsys_response_factory()
+        redsys_response["Ds_Response"] = "190"  # denied
+        merchant_params = encode_response(redsys_response)
+        signature = compute_signature(
+            str(redsys_response["Ds_Order"]),
+            merchant_params,
+            self.redsys.shared_secret,
+        )
+
+        request = self.factory.get(
+            "/",
+            data={
+                "Ds_SignatureVersion": "HMAC_SHA256_V1",
+                "Ds_MerchantParameters": merchant_params,
+                "Ds_Signature": signature,
+            },
+        )
+
+        result = self.redsys.process_data(self.payment, request)
+
+        assert result.status_code == 302
+        assert result.url == f"http://localhost:8000/{self.payment.pk}/failure"
+        self.payment.refresh_from_db()
+        assert self.payment.status == "rejected"
+
+    @patch("payments_redsys.compare_signatures", Mock(return_value=True))
+    @patch("payments_redsys.requests.post")
+    def test_refund_mock(self, post: MagicMock):
+        post.return_value.content = json.dumps(
+            {
+                "Ds_SignatureVersion": "HMAC_SHA256_V1",
+                "Ds_MerchantParameters": base64.b64encode(
+                    json.dumps({"Ds_Response": "0400"}).encode()
+                ).decode(),
+                "Ds_Signature": "...",
+            }
+        ).encode("utf-8")
+        refund_amount = Decimal("5.0")
+
+        result = self.redsys.refund(self.payment, refund_amount)
+
+        assert result == refund_amount
+        post.assert_called_once_with(
+            "https://sis-t.redsys.es:25443/sis/rest/trataPeticionREST",
+            json={
+                "Ds_SignatureVersion": "HMAC_SHA256_V1",
+                "Ds_MerchantParameters": "eyJEU19NRVJDSEFOVF9BTU9VTlQiOiAiNTAwIiwgIkRTX01FUkNIQU5UX0NVUlJFTkNZIjogIjk3OCIsICJEU19NRVJDSEFOVF9NRVJDSEFOVENPREUiOiAiOTk5MDA4ODgxIiwgIkRTX01FUkNIQU5UX09SREVSIjogIlJFRFNZUz1URVNUOjAwMDAwMSIsICJEU19NRVJDSEFOVF9URVJNSU5BTCI6ICIwMDEiLCAiRFNfTUVSQ0hBTlRfVFJBTlNBQ1RJT05UWVBFIjogIjMifQ==",
+                "Ds_Signature": "Lp4Akv6gg+IDsAwKC4Kc1+lob5dfjST11GkFISpSkgQ=",
+            },
+        )
+
+    @pytest.mark.skip("Can't make the refund test endpoint return success")
```

Comment:
I'm wondering how you tested refunds. I haven't yet managed a proper one with the Redsys test enviroment. The call goes fine but it complains the refund amount is too high... ?

Reply:
I managed to test the refund, altough with a non-automatable process (manually feeding it a valid order number). It works.